File: web/src/lib/terminal.ts
import { Terminal } from "@xterm/xterm";
import { FitAddon } from "@xterm/addon-fit";
import { socketURL } from "./api";

/**
 * The terminal socket is deliberately not a JSON protocol.
 *
 * Binary messages are terminal bytes in both directions: the daemon's
 * exact state first, then live output, with keystrokes travelling back
 * the same way. Text messages are the control channel — a size, or a
 * notice that what follows is state rather than output — and there are
 * few of them.
 */
interface Control {
  type: "resize" | "seed";
  cols?: number;
  rows?: number;
}

/**
 * The smallest geometry worth telling the daemon about. Well below any
 * usable pane and well above what a collapsed one reports.
 */
const minCols = 20;
const minRows = 4;

/** How long to wait before reattaching, and the ceiling it climbs to. */
const retryFloor = 500;
const retryCeiling = 10_000;

/**
 * The backstop for the batch below, in milliseconds.
 *
 * A frame is the cadence that matters, and in a tab being painted the
 * frame always wins this — which is the point of the gap. A hidden tab
 * is served no frames at all, though, and a replica that simply stopped
 * applying output would be stale when the tab came back, having held
 * every unapplied byte until then. The timer is what drains that tab;
 * the browser throttles it to about a second, which bounds what a hidden
 * tab holds without ever pacing a visible one.
 */
const batchBackstop = 250;

/**
 * RIS, as bytes, so it batches with the output around it rather than
 * breaking the run into two writes.
 */
const reset = new Uint8Array([0x1b, 0x63]);

/**
 * DECSET 2026, begin and end: the terminal is told to hold its paint
 * until the whole thing has landed.
 *
 * Rebuilding a replica is a reset followed by a screen and everything
 * that scrolled off above it, and that is far more than a parser gets
 * through before it yields to let the renderer run. Painted halfway, the
 * reset has happened and the history has not: an empty buffer, the view
 * at its first line, and a pane that appears to leap to the top and back
 * a moment later. Wrapped, there is no halfway to paint — the same
 * mechanism the agents themselves use to keep a frame from being seen
 * half drawn.
 */
const beginSync = new Uint8Array([0x1b, 0x5b, 0x3f, 0x32, 0x30, 0x32, 0x36, 0x68]);
const endSync = new Uint8Array([0x1b, 0x5b, 0x3f, 0x32, 0x30, 0x32, 0x36, 0x6c]);

/** rebuild is state to replace a replica with, made unpaintable-halfway. */
function rebuild(state: Uint8Array): Uint8Array {
  const whole = new Uint8Array(
    beginSync.length + reset.length + state.length + endSync.length,
  );
  let at = 0;
  for (const part of [beginSync, reset, state, endSync]) {
    whole.set(part, at);
    at += part.length;
  }
  return whole;
}

/**
 * Whether this page batches at all.
 *
 * `?batch=off` writes every message straight through the way it did
 * before batching existed, so the two can be compared by reloading. The
 * batch buys atomicity and costs a frame of latency — output waits for a
 * frame to be handed over and then waits for another to be painted —
 * and which of those matters more is a question about a real agent on a
 * real screen rather than one to settle by reasoning.
 */
function batching(): boolean {
  try {
    return new URL(window.location.href).searchParams.get("batch") !== "off";
  } catch {
    return true;
  }
}

/**
 * Whether this page narrates what the socket is doing.
 *
 * A replica rebuilt is a screen that jumps, and the two reasons — the
 * daemon resyncing a viewer that fell behind, and the socket coming back
 * after a drop — look identical from the buffer's side. They do not look
 * identical from here.
 */
function narrating(): boolean {
  try {
    return new URL(window.location.href).searchParams.get("probe") !== "0";
  } catch {
    return false;
  }
}

function usable(cols: number, rows: number): boolean {
  return cols >= minCols && rows >= minRows;
}

export type Connection = "live" | "reconnecting";

export interface Attachment {
  /** Re-measures and tells the daemon; call on layout changes. */
  fit(): void;
  close(): void;
}

export interface Options {
  /**
   * watching attaches without ever speaking: no keystrokes, and no size.
   *
   * The wall is why this exists. A terminal belongs to its session and
   * every viewer shares it, so a grid of small cells that each announced
   * their geometry would reflow the whole fleet to the size of the
   * smallest tile — and the dashboard reading one of those agents would
   * watch it happen. A watcher renders whatever size the terminal
   * already is, which the seed tells it.
   */
  watching?: boolean;
}

/**
 * attach keeps one xterm.js instance bound to one agent's terminal, across
 * however many sockets that takes.
 *
 * The subtlety worth stating: a `seed` notice means the binary message
 * behind it *replaces* the replica rather than extending it. It arrives at
 * attach, and again whenever this viewer fell behind and the daemon sent
 * state instead of the output it missed. That is also what makes
 * reconnecting safe — a new socket opens with a seed, so the replica is
 * rebuilt rather than continued from wherever it was cut off.
 */
export function attach(
  term: Terminal,
  fitAddon: FitAddon,
  id: string,
  isLaidOut: () => boolean,
  onConnection: (state: Connection) => void = () => {},
  options: Options = {},
): Attachment {
  const watching = options.watching ?? false;
  let socket: WebSocket | null = null;
  // Zero until this viewer has measured itself: it has asserted no
  // geometry, so it has none to compare against.
  let announced = { cols: 0, rows: 0 };
  let retry = retryFloor;
  let retryTimer: number | undefined;
  let closed = false;

  /**
   * Output reaches xterm one batch per frame, never one write per
   * message.
   *
   * A busy agent sends output in hundreds of small messages a second,
   * and handing each one to term.write() from the message handler
   * starves the renderer: the replica keeps parsing — its buffer stays
   * current — while the screen stops repainting entirely, then snaps
   * forward when the agent goes quiet. That blank-and-catch-up is the
   * flash you see while an agent is working, and it is a scheduling
   * problem rather than a throughput one: the same bytes, delivered in
   * one batch per frame, paint continuously.
   *
   * A batch is a queue rather than a buffer because not everything in it
   * is bytes. A resize has to land behind the output it follows (see the
   * resize notice below), so it rides the same queue as a callback on
   * the write in front of it — which is what keeps a repaint from being
   * reflowed by a size that belongs after it.
   */
  const batched = batching();
  const narrate = narrating();
  let seeds = 0;
  let sockets = 0;
  let batch: Array<Uint8Array | (() => void)> = [];
  let frame: number | undefined;
  let backstop: number | undefined;

  const flush = () => {
    if (frame !== undefined) window.cancelAnimationFrame?.(frame);
    window.clearTimeout(backstop);
    frame = backstop = undefined;
    const queued = batch;
    batch = [];
    let run: Uint8Array[] = [];
    let size = 0;
    // One write per run of bytes, and the run ends where an action has to
    // be ordered against it.
    const writeRun = (done?: () => void) => {
      if (size === 0) {
        if (done) term.write("", done);
        return;
      }
      const joined = new Uint8Array(size);
      let at = 0;
      for (const chunk of run) {
        joined.set(chunk, at);
        at += chunk.length;
      }
      run = [];
      size = 0;
      term.write(joined, done);
    };
    for (const item of queued) {
      if (typeof item === "function") {
        writeRun(item);
        continue;
      }
      run.push(item);
      size += item.length;
    }
    writeRun();
  };

  const enqueue = (item: Uint8Array | (() => void)) => {
    if (!batched) {
      if (typeof item === "function") term.write("", item);
      else term.write(item);
      return;
    }
    batch.push(item);
    if (frame !== undefined || backstop !== undefined) return;
    frame = window.requestAnimationFrame?.(flush);
    backstop = window.setTimeout(flush, batchBackstop);
  };

  const open = () => {
    if (closed) return;
    // A pane that has not been laid out has no geometry worth asserting,
    // and this terminal is shared: any size named here moves the agent's
    // real terminal for every viewer, the dashboard included, and nothing
    // moves it back. Naming none leaves it where it is until this pane
    // has a shape and fit() can speak for it (#155).
    if (!watching && isLaidOut()) {
      // Re-measure rather than trusting term.cols, which may hold a size
      // the daemon pushed rather than one this pane has. A watcher skips
      // it: its grid is whatever the terminal already is, and measuring
      // its own tile would only invite it to say so.
      fitAddon.fit();
    }
    const measured = !watching && isLaidOut() && usable(term.cols, term.rows);
    const opening = measured
      ? { cols: term.cols, rows: term.rows }
      : { cols: 0, rows: 0 };
    announced = opening;

    sockets++;
    if (narrate) {
      console.log(
        `flicker: opening terminal socket #${sockets}` +
          (sockets > 1 ? " — THE SOCKET DROPPED AND CAME BACK" : ""),
      );
    }
    const live = new WebSocket(
      socketURL(
        `/api/agents/${id}/terminal`,
        measured
          ? { cols: String(opening.cols), rows: String(opening.rows) }
          : {},
      ),
    );
    socket = live;
    live.binaryType = "arraybuffer";

    // Set by a seed notice and consumed by the binary message behind it.
    // The daemon pairs them one for one; a stray one degrades to a
    // repaint rather than to a screen that keeps state it was told to
    // drop.
    let replacing = false;

    live.onopen = () => {
      onConnection("live");
      // Measure now: the pane may have been laid out since this viewer
      // last spoke, and the size in the URL was read before that.
      fit();
    };

    live.onmessage = (event) => {
      if (typeof event.data === "string") {
        const control: Control = JSON.parse(event.data);
        if (control.type === "seed") {
          seeds++;
          if (narrate) {
            console.log(
              `flicker: the daemon seeded this replica (#${seeds} on socket ` +
                `#${sockets})` +
                (seeds > 1
                  ? " — A RESYNC: this viewer fell behind and was handed state"
                  : ""),
            );
          }
          // The attach worked. Not `onopen`: this server upgrades before
          // it attaches, so a failed attach is an accepted socket that
          // closes a moment later — and treating that as success resets
          // the backoff, turning an unreachable daemon into two full
          // attach attempts a second, forever.
          retry = retryFloor;
          replacing = true;
          return;
        }
        if (control.type === "resize" && control.cols && control.rows) {
          const { cols, rows } = control;
          // Behind the write queue, not in front of it. term.resize is
          // synchronous while term.write is buffered, so resizing here
          // would reflow output that arrived before this notice — the
          // repaint that belongs to this size is queued behind it.
          // Deferred, so it lands behind the output it belongs to — and
          // guarded, because the queue it waits in is not drained by
          // dispose(): switching agents inside that window would resize a
          // terminal that no longer exists.
          enqueue(() => {
            if (closed) return;
            term.resize(cols, rows);
          });
          // The daemon just told us the size, so it is ours now. Without
          // this the next fit() measures the same grid, compares it to a
          // stale record, and quietly snaps the replica back to a size
          // nobody else is using.
          //
          // This is also how a viewer that asserted nothing learns the
          // geometry its seed was wrapped for: the size arrives just
          // ahead of the state it belongs to.
          announced = { cols, rows };
        }
        return;
      }
      const bytes = new Uint8Array(event.data as ArrayBuffer);
      if (replacing) {
        replacing = false;
        // The reset rides with the state it precedes, in one write, held
        // behind a synchronized update.
        //
        // Through the write queue rather than term.reset(), which is
        // synchronous: bytes already handed to write() but not yet parsed
        // would land on the fresh screen after the reset — the doubled
        // screen this exists to prevent, on the one path where there is
        // certain to be output in flight. And wrapped, because a replica
        // is rebuilt often enough — every attach, and every resync of a
        // viewer that fell behind — that a half-drawn one is the flicker
        // rather than a curiosity.
        enqueue(rebuild(bytes));
        return;
      }
      enqueue(bytes);
    };

    // The socket ends for reasons this side cannot tell apart: the agent
    // exited, the server restarted, the laptop slept. Reattaching answers
    // all of them — a live agent seeds a fresh replica, and a gone one
    // disappears from the roster, which takes this pane with it.
    live.onclose = () => {
      if (closed || socket !== live) return;
      onConnection("reconnecting");
      retryTimer = window.setTimeout(open, retry);
      retry = Math.min(retry * 2, retryCeiling);
    };
  };

  const send = (data: Uint8Array<ArrayBuffer> | string) => {
    if (socket?.readyState === WebSocket.OPEN) socket.send(data);
  };

  // A watcher registers no input handlers at all, rather than dropping
  // what they produce: a cell that quietly swallowed keystrokes would
  // look like an agent ignoring you.
  const input = watching
    ? { dispose: () => {} }
    : term.onData((data) => {
        const encoded = new TextEncoder().encode(data);
        const bytes = new Uint8Array(new ArrayBuffer(encoded.length));
        bytes.set(encoded);
        send(bytes);
      });
  const binary = watching
    ? { dispose: () => {} }
    : term.onBinary((data) => {
        const bytes = new Uint8Array(new ArrayBuffer(data.length));
        for (let i = 0; i < data.length; i++) {
          bytes[i] = data.charCodeAt(i) & 255;
        }
        send(bytes);
      });

  const fit = () => {
    // A watcher never claims a size; it renders the one it was given.
    if (watching) return;
    // A collapsed or hidden pane has no opinion about geometry, and this
    // terminal is shared.
    if (!isLaidOut()) return;
    fitAddon.fit();
    if (!usable(term.cols, term.rows)) return;
    if (term.cols === announced.cols && term.rows === announced.rows) return;
    announced = { cols: term.cols, rows: term.rows };
    send(JSON.stringify({ type: "resize", cols: term.cols, rows: term.rows }));
  };

  open();

  return {
    fit,
    close() {
      closed = true;
      window.clearTimeout(retryTimer);
      // Nothing queued outlives the terminal it was queued for: the
      // batch is dropped rather than flushed, because the flush would
      // write into an emulator this pane is about to dispose.
      if (frame !== undefined) window.cancelAnimationFrame?.(frame);
      window.clearTimeout(backstop);
      frame = backstop = undefined;
      batch = [];
      input.dispose();
      binary.dispose();
      if (socket) {
        // Detach before closing: a socket in CLOSING can still dispatch a
        // buffered message, and the terminal it would write to is about
        // to be disposed.
        socket.onmessage = null;
        socket.onclose = null;
        socket.close();
      }
    },
  };
}

File: internal/api/terminal.go
package api

import (
	"context"
	"encoding/json"

	"net/http"
	"strconv"
	"time"

	"github.com/coder/websocket"

	"github.com/trentkm/stormlight/internal/diagnostic"
	"github.com/trentkm/stormlight/internal/pty"
)

// The terminal socket is deliberately not a JSON protocol.
//
// Binary messages are terminal bytes, raw and unwrapped, in both
// directions: the daemon's exact snapshot arrives first, then live
// output; keystrokes travel back the same way. Wrapping a keypress in
// JSON would put an encode, a decode and an allocation on the path
// between a key and the process that answers it, for no gain — the frame
// already carries the length.
//
// Text messages are the control channel: a size, or a notice that the
// terminal moved under this viewer. There are few of them and they are
// never in the typing path.
type controlMessage struct {
	Type string `json:"type"`
	Cols int    `json:"cols,omitempty"`
	Rows int    `json:"rows,omitempty"`
}

const (
	controlResize = "resize"
	// controlSeed precedes a binary message that replaces the replica
	// rather than extending it: the first one at attach, and any later
	// one when the daemon resynced a viewer that fell behind.
	controlSeed = "seed"
)

// terminal relays one agent's live terminal over one WebSocket.
func (s *Server) terminal(w http.ResponseWriter, r *http.Request) {
	cols, rows := terminalSize(r)

	// Upgrade before attaching. Attaching resizes the agent's terminal —
	// the daemon owns one terminal per agent and every viewer shares it —
	// so doing it first would let a plain GET, which cannot become a
	// terminal at all, reflow the pane a dashboard user is reading.
	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{
		// Origin is checked in the auth middleware, against the loopback
		// hosts this server actually serves.
		InsecureSkipVerify: true,
	})
	if err != nil {
		return
	}
	transport, err := s.service.AttachTerminal(r.Context(), r.PathValue("id"), cols, rows)
	if err != nil {
		_ = conn.Close(websocket.StatusInternalError, "attach failed")
		return
	}
	// A paste is one message, and terminals receive large ones. The
	// library's default read limit is 32 KiB, which would not reject the
	// paste but tear down the whole attachment.
	conn.SetReadLimit(maxTerminalMessage)
	// A terminal is a long conversation; nothing about it should time out
	// for being quiet, so the relay runs on a context of its own that
	// ends when either side hangs up.
	ctx, cancel := context.WithCancel(context.WithoutCancel(r.Context()))
	defer cancel()
	defer transport.Close()
	defer conn.CloseNow()

	// The daemon's snapshot is state, not output: it replaces whatever the
	// replica had. Say so before sending it, so a reconnecting client
	// never appends a screen to the one it was already showing — and send
	// the size it was rendered at first, because a client that named no
	// geometry of its own has no other way to learn what these bytes were
	// wrapped for.
	if err := writeState(ctx, conn, transport.Seed()); err != nil {
		return
	}

	go s.pumpInput(ctx, cancel, conn, transport)
	go keepalive(ctx, cancel, conn)

	// Output goes out in bursts rather than in reads.
	//
	// A repaint is not one read. The PTY hands back whatever its buffer
	// held when the reader got there, so a program that erases the
	// screen and redraws it in a single breath still arrives split —
	// measured on this socket, the message carrying codex's ESC[2J and
	// the message carrying the redraw behind it land a tenth of a
	// millisecond apart. Relayed separately, a viewer can paint between
	// them, and what it paints is the erased screen: the pane blanks and
	// comes back.
	//
	// Nothing downstream can repair that. A client writes a message into
	// its replica indivisibly, so bytes that share a message are painted
	// together and bytes that do not, may not be. The atomicity has to
	// be built here, where the pieces are still adjacent: everything
	// already queued leaves together, and a brief pause lets the rest of
	// a repaint catch up. It is what a local terminal gets for free by
	// draining its fd before it renders.
	for {
		select {
		case <-ctx.Done():
			return
		case message, ok := <-transport.Output():
			if !ok {
				// The stream ended — and this layer cannot say why. The
				// agent may have exited, or the attachment may have ended
				// some other way. Saying "session ended" here would have
				// a live agent read as a dead one; the client reconnects
				// and asks the roster, which knows. Falling behind is no
				// longer among the reasons: this attachment resyncs.
				_ = conn.Close(websocket.StatusNormalClosure, "stream ended")
				return
			}
			for pending := &message; pending != nil; {
				next, err := relay(ctx, conn, transport.Output(), *pending)
				if err != nil {
					return
				}
				pending = next
			}
		}
	}
}

// relay writes one delivery and returns whatever ended it, so the caller
// can write that in turn. Ordering is the contract this whole plane rests
// on, and this is where it is kept: state and sizes go out on their own,
// output goes out as the burst it belongs to, and a notice that arrived
// mid-burst waits for the bytes in front of it.
func relay(
	ctx context.Context,
	conn *websocket.Conn,
	output <-chan pty.Message,
	message pty.Message,
) (*pty.Message, error) {
	switch {
	case message.Resync != nil:
		// State first: a resync carries its own size, and the resize
		// branch below would otherwise swallow it.
		//
		// This viewer fell behind and the daemon sent state instead of
		// the bytes it missed — the same shape as the opening seed, and
		// written by the same function, because they are the same thing
		// at different moments.
		return nil, writeState(ctx, conn, message)
	case message.Resize != nil:
		// A terminal this viewer shares can be moved by anyone holding
		// it — the dashboard, a full-screen attach — and a replica
		// painting at a size nobody else is using is a diverged replica.
		// The notice travels just ahead of the repaint that belongs to
		// it.
		return nil, writeControl(ctx, conn, controlMessage{
			Type: controlResize,
			Cols: message.Resize.Cols,
			Rows: message.Resize.Rows,
		})
	}
	burst, next := gatherBurst(ctx, output, message.Bytes)
	if len(burst) == 0 {
		return next, nil
	}
	if err := conn.Write(ctx, websocket.MessageBinary, burst); err != nil {
		return nil, err
	}
	return next, nil
}

const (
	// burstQuiet is how long output waits for the rest of its repaint.
	// The pieces of one arrive a fraction of a millisecond apart, so this
	// is far longer than the gap inside a repaint and far shorter than
	// the gap a person can see.
	burstQuiet = 3 * time.Millisecond
	// burstCap bounds that wait, so a stream that never falls quiet
	// reaches the viewer on time rather than being held for a pause that
	// is not coming.
	burstCap = 16 * time.Millisecond
	// burstBytes bounds one message. A repaint is kilobytes; this is the
	// backstop for a program dumping a file, where atomicity buys nothing
	// and holding megabytes costs.
	burstBytes = 256 << 10
)

// gatherBurst collects consecutive output into one message, stopping at
// the first pause, at the cap, or at a notice that has to be written on
// its own — which it hands back rather than swallowing.
func gatherBurst(
	ctx context.Context,
	output <-chan pty.Message,
	first []byte,
) (burst []byte, next *pty.Message) {
	burst = append(make([]byte, 0, len(first)), first...)
	quiet := time.NewTimer(burstQuiet)
	defer quiet.Stop()
	deadline := time.NewTimer(burstCap)
	defer deadline.Stop()
	for len(burst) < burstBytes {
		select {
		case <-ctx.Done():
			return burst, nil
		case message, ok := <-output:
			if !ok {
				return burst, nil
			}
			if message.Resync != nil || message.Resize != nil {
				held := message
				return burst, &held
			}
			burst = append(burst, message.Bytes...)
			// Stopped before reset, and drained if it had already fired:
			// a timer reset with its value still in the channel fires at
			// once, which would end every burst at its first chunk and
			// put this right back where it started.
			if !quiet.Stop() {
				<-quiet.C
			}
			quiet.Reset(burstQuiet)
		case <-quiet.C:
			return burst, nil
		case <-deadline.C:
			return burst, nil
		}
	}
	return burst, nil
}

// pumpInput carries the client's keystrokes and resizes to the terminal.
func (s *Server) pumpInput(
	ctx context.Context,
	cancel context.CancelFunc,
	conn *websocket.Conn,
	transport pty.Transport,
) {
	defer cancel()
	for {
		kind, payload, err := conn.Read(ctx)
		if err != nil {
			return
		}
		switch kind {
		case websocket.MessageBinary:
			if err := transport.Write(payload); err != nil {
				return
			}
		case websocket.MessageText:
			var message controlMessage
			if err := json.Unmarshal(payload, &message); err != nil {
				continue
			}
			if message.Type != controlResize {
				continue
			}
			// Refuse a size rather than correct it. Clamping upward looks
			// harmless and is not: a message with no size at all would
			// become a legal 2x2, and one resize reflows the terminal
			// every viewer shares — including the dashboard someone is
			// reading. An unusable number is a client bug; the terminal
			// is not the place to absorb it.
			if !usableSize(message.Cols, message.Rows) {
				continue
			}
			if err := transport.Resize(ctx, message.Cols, message.Rows); err != nil {
				// A resize that loses a race with the session ending is
				// not worth dropping the connection over.
				diagnostic.Logger().Debug("terminal resize failed", "error", err)
			}
		}
	}
}

// keepaliveInterval paces the ping that proves a viewer is still there.
// A laptop that sleeps or a wifi drop leaves a connection that looks open
// and never speaks again; without this the server holds the goroutine and
// the daemon attachment behind it until the OS gives up on the socket,
// which can be hours.
const keepaliveInterval = 30 * time.Second

func keepalive(ctx context.Context, cancel context.CancelFunc, conn *websocket.Conn) {
	defer cancel()
	ticker := time.NewTicker(keepaliveInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
		}
		pingCtx, timeout := context.WithTimeout(ctx, keepaliveInterval)
		err := conn.Ping(pingCtx)
		timeout()
		if err != nil {
			return
		}
	}
}

// writeState sends terminal state — the attach seed, or a resync — as the
// size it was rendered at, then the notice, then the bytes.
//
// The size leads because the client has to size the replica it is about
// to fill; the notice is what tells it these bytes replace rather than
// extend.
func writeState(ctx context.Context, conn *websocket.Conn, state pty.Message) error {
	if state.Resize != nil {
		if err := writeControl(ctx, conn, controlMessage{
			Type: controlResize,
			Cols: state.Resize.Cols,
			Rows: state.Resize.Rows,
		}); err != nil {
			return err
		}
	}
	if err := writeControl(ctx, conn, controlMessage{Type: controlSeed}); err != nil {
		return err
	}
	return conn.Write(ctx, websocket.MessageBinary, state.Resync)
}

func writeControl(ctx context.Context, conn *websocket.Conn, message controlMessage) error {
	payload, err := json.Marshal(message)
	if err != nil {
		return err
	}
	return conn.Write(ctx, websocket.MessageText, payload)
}

// A terminal's geometry is not a matter of opinion, and this is the one
// place a client's number reaches the daemon's emulator. The emulator
// allocates the whole grid — screen and scrollback both — so an unbounded
// size is an out-of-memory request served on demand, and the daemon that
// serves it owns every agent's process.
//
// The bounds are sized against real displays with room to spare, not
// against what the format allows: an ultrawide monitor at a small font
// reaches perhaps 400 columns, and rows past a couple of hundred are a
// scrollback question rather than a screen one. Generous limits are not
// free here — scrollback is columns times its line count, so width is
// what actually costs, and a cap ten times any real terminal would mean
// hundreds of megabytes per session for no one's benefit.
const (
	maxTerminalCols = 500
	maxTerminalRows = 200
)

// usableSize reports whether a client's geometry is one a terminal can
// actually be.
func usableSize(cols, rows int) bool {
	return cols >= 2 && rows >= 2 && cols <= maxTerminalCols && rows <= maxTerminalRows
}

// maxTerminalMessage bounds one inbound message — a keystroke, or a
// paste. Generous enough for pasting a file, small enough that a client
// cannot make the server hold an arbitrary buffer.
const maxTerminalMessage = 4 << 20

// terminalSize reads the viewer's geometry from the query string, or
// zero when it names none.
//
// Zero means "no opinion", and it is not the same as a default. The
// terminal is shared, so any size this returns is asserted on the agent
// for every viewer — a browser tab that has not laid itself out yet would
// otherwise reflow the dashboard's pane to 80x24 and leave it there,
// because nothing re-asserts the dashboard's own size afterward. A viewer
// that cannot say has nothing to say; it will resize as soon as it has a
// shape (see #155).
func terminalSize(r *http.Request) (cols, rows int) {
	cols, rows = queryInt(r, "cols"), queryInt(r, "rows")
	if !usableSize(cols, rows) {
		return 0, 0
	}
	return cols, rows
}

func queryInt(r *http.Request, name string) int {
	value, err := strconv.Atoi(r.URL.Query().Get(name))
	if err != nil {
		return 0
	}
	return value
}
